### PR TITLE
Added GridGenerator::generate_from_name_and_arguments

### PR DESCRIPTION
**Pull request by luca-heltai**
Prequel to ParsedGridGenerator.

### CODE REVIEW

**Comment by luca-heltai** [comment]
> Did you rebase the pull request on top of a recent version of the master branch? We should already include the correct header...

yes. This is rebased on the current master, after the merge of the mutable bind. The header is definitely there. This seems to me like the compiler advertise itself as cxx17 compliant, but then fails to deliver `std::apply`. Is this possible?

**Comment by luca-heltai** [comment]
I think I might have fixed the issue. This was not presenting itself in the other PR because we were including `tuple.h` anyway in another file down the line.

**Comment by luca-heltai** [comment]
I still see the same problem on the mac build. @tjhei , is `--stdlib=libc++` on for the mac build? The compiler advertises itself as cxx17 compliant, but then `std::apply` is not there. It may be that it is using gcc libs instead of clang ones.

**Comment by luca-heltai** [comment]
Help!

I don't know how to solve this:

```
FAILED: /usr/bin/mpicxx   -DBOOST_NO_AUTO_PTR -DDEBUG -DTBB_DO_ASSERT=1 -DTBB_USE_DEBUG -Isource/grid -Iinclude -I/jenkins/workspace/dealii_PR-8654/include -I/jenkins/workspace/dealii_PR-8654/bundled/tbb-2018_U2/include -I/jenkins/workspace/dealii_PR-8654/bundled/boost-1.70.0/include -I/jenkins/workspace/dealii_PR-8654/bundled/umfpack/UMFPACK/Include -I/jenkins/workspace/dealii_PR-8654/bundled/umfpack/AMD/Include -I/jenkins/workspace/dealii_PR-8654/bundled/muparser_v2_2_4/include -isystem /usr/include/mpich -isystem /home/dealii/libs/parmetis-4.0.3/include -isystem /home/dealii/libs/trilinos-release-12-10-1/include -isystem /home/dealii/libs/superlu_dist_5.1.2/include -isystem /home/dealii/libs/petsc-3.7.6/include -isystem /home/dealii/libs/adolc-2.6.4-rc1/include -isystem /home/dealii/libs/assimp-4.1.0/include -isystem /home/dealii/libs/hdf5-1.10.1/include -isystem /home/dealii/libs/nanoflann-v1.2.3/include -isystem /home/dealii/libs/oce-OCE-0.18.2/include/oce -isystem /home/dealii/libs/p4est-2.0/FAST/include -isystem /home/dealii/libs/slepc-3.7.3/include -isystem /home/dealii/libs/sundials-3.1.0/include -pedantic -fPIC -Wall -Wextra -Woverloaded-virtual -Wpointer-arith -Wsign-compare -Wsuggest-override -Wswitch -Wsynth -Wwrite-strings -Wno-deprecated-declarations -Wno-literal-suffix -Wno-psabi -fopenmp-simd -std=c++14 -std=c++14 -Werror -Wno-parentheses -Wno-unused-local-typedefs -O0 -ggdb -Wa,--compress-debug-sections -MMD -MT source/grid/CMakeFiles/obj_grid_debug.dir/grid_generator_from_name.cc.o -MF source/grid/CMakeFiles/obj_grid_debug.dir/grid_generator_from_name.cc.o.d -o source/grid/CMakeFiles/obj_grid_debug.dir/grid_generator_from_name.cc.o -c /jenkins/workspace/dealii_PR-8654/source/grid/grid_generator_from_name.cc

/jenkins/workspace/dealii_PR-8654/source/grid/grid_generator_from_name.cc: In function ‘bool dealii::GridGenerator::{anonymous}::generate_special(const string&, const string&, dealii::Triangulation<3, 3>&)’:

/jenkins/workspace/dealii_PR-8654/source/grid/grid_generator_from_name.cc:274:64: error: no matching function for call to ‘parse_and_create(void (&)(dealii::Triangulation<3, 3>&, unsigned int, unsigned int, double, double), const string&, dealii::Triangulation<3, 3>&)’

                          const double>(moebius, arguments, tria);

                                                                ^

/jenkins/workspace/dealii_PR-8654/source/grid/grid_generator_from_name.cc:28:5: note: candidate: template<int dim, int spacedim, class ... Arguments> void dealii::GridGenerator::{anonymous}::parse_and_create(void (*)(dealii::Triangulation<dim, spacedim>&, Arguments ...), const string&, dealii::Triangulation<dim, spacedim>&)

     parse_and_create(void (*generator)(Triangulation<dim, spacedim> &,

     ^

/jenkins/workspace/dealii_PR-8654/source/grid/grid_generator_from_name.cc:28:5: note:   template argument deduction/substitution failed:

/jenkins/workspace/dealii_PR-8654/source/grid/grid_generator_from_name.cc:274:64: note:   mismatched types ‘const unsigned int’ and ‘unsigned int’

                          const double>(moebius, arguments, tria);

                                                                ^

ninja: build stopped: subcommand failed.

script returned exit code 1
```
In the function call, I'm using the correct argument list. For some reason I ignore, the compiler is ignoring the const qualifiers for the moebius function. All other function calls are done correctly (and using the same scheme). The only difference I see here, is that the moebius function is not templated, while all others are. Is there a reason why this should fail?

```
if (name == "moebius")
        parse_and_create<3,
                         3,
                         const unsigned int,
                         const unsigned int,
                         const double,
                         const double>(moebius, arguments, tria);
```
The compiler complains about not being able to match `parse_and_create(void (&)(dealii::Triangulation<3, 3>&, unsigned int, unsigned int, double, double), const string&, dealii::Triangulation<3, 3>&)`, but I'm *not* calling it with ` unsigned int, unsigned int, double, double`... I'm lost here.

**Comment by luca-heltai** [comment]
Thanks @masterleinad !

**Comment by luca-heltai** [comment]
Anyone else?

File: source/grid/grid_generator_from_name.cc [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+    }
+  } // namespace
+
+  template <int dim, int spacedim>
+  void
+  generate_from_name_and_arguments(Triangulation<dim, spacedim> &tria,
+                                   const std::string &           name,
+                                   const std::string &           arguments)
+  {
+    {
```
Comment (masterleinad) [inline source/grid/grid_generator_from_name.cc]:
```suggestion
```

File: source/grid/grid_generator_from_name.cc [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+      else if (generate_codimension_zero_grid(name, arguments, tria))
+        {}
+      else if (generate_codimension_one_grid(name, arguments, tria))
+        {}
+      else if (generate_special(name, arguments, tria))
+        {}
+      else
+        AssertThrow(false,
+                    ExcMessage(name + "(" + arguments + ") not implemented"));
+    }
```
Comment (masterleinad) [inline source/grid/grid_generator_from_name.cc]:
```suggestion
```

**Comment by luca-heltai** [comment]
@masterleinad, done.

File: source/grid/grid_generator_from_name.cc [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+              parse_and_create<dim,
+                               dim,
+                               const std::vector<unsigned int> &,
+                               const Point<dim> &,
+                               const Point<dim> &,
+                               bool>(subdivided_hyper_rectangle,
+                                     arguments,
+                                     tria);
+            }
+          catch (...)
```
Comment (tjhei) [inline source/grid/grid_generator_from_name.cc]: Can you please document why this is needed? Would it make sense to only catch the specific exception happening here?

File: source/grid/grid_generator_from_name.cc [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+                                                             tria);
+      else
+        {
+          return false;
+        }
+      return true;
+    }
+  } // namespace
+
+  template <int dim, int spacedim>
```
Comment (bangerth) [inline source/grid/grid_generator_from_name.cc]:
```suggestion


  template <int dim, int spacedim>
```

File: source/grid/grid_generator_from_name.cc [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+      parse_and_create<dim,
+                       spacedim,
+                       const std::vector<Point<spacedim>> &,
+                       bool>(general_cell, arguments, tria);
+    else if (name == "hyper_cross")
+      parse_and_create<dim, spacedim, const std::vector<unsigned int> &, bool>(
+        hyper_cross, arguments, tria);
+    else if (generate_codimension_zero_grid(name, arguments, tria))
+      {}
+    else if (generate_codimension_one_grid(name, arguments, tria))
```
Comment (bangerth) [inline source/grid/grid_generator_from_name.cc]: Ha, funny sequence here. I think a comment of the form "Now try each of the following options one after the other and see which one works" would be helpful.

File: include/deal.II/grid/grid_generator.h [hunk 6]
```diff
@@ -1328,6 +1328,47 @@ namespace GridGenerator
...
+   * Patterns::Tools::Convert<TupleTyple>::to_value(), where `TupleType` here is
+   * a tuple containing **all** the arguments of the GridGenerator function,
+   * including all optional arguments.
+   *
+   * An example usage of this function is given by:
+   * @code
+   * GridGenerator::generate_from_name_and_arguments(
+   *   tria,
+   *   "hyper_ball",
+   *   "0.0, 0.0: 1: false");
```
Comment (bangerth) [inline include/deal.II/grid/grid_generator.h]:
```suggestion
   *   "0.0, 0.0 : 1 : false");
```

File: include/deal.II/grid/grid_generator.h [hunk 7]
```diff
@@ -1328,6 +1328,47 @@ namespace GridGenerator
...
+   * @endcode
```
Comment (bangerth) [inline include/deal.II/grid/grid_generator.h]:
```suggestion
   * @endcode
   * Here, the colon separates the function arguments, and the comma separates the coordinates of a 
   * Point<2> argument.
```

File: source/grid/grid_generator_from_name.cc [hunk 8]
```diff
@@ -0,0 +1,331 @@
...
+                         double,
+                         double,
+                         unsigned int,
+                         bool>(half_hyper_shell, arguments, tria);
+
+
+      else if (name == "quarter_hyper_shell")
+        parse_and_create<dim,
+                         dim,
+                         const Point<dim> &,
```
[the remark this answers is not in the document]
Comment (luca-heltai) [reply source/grid/grid_generator_from_name.cc]: Apparently, not all compilers are good at doing this. Without explicitly stating the arguments, gcc compiles, clang 6 compiles, clang 7 and 8 don't. I'm not sure if we have a smart way (something like `Identity<type>::type`, that we used in the past). I tried to get this right on the three machines I hade, and this was the faster solution in the end...

[the remark this answers is not in the document]
Comment (bangerth) [reply source/grid/grid_generator_from_name.cc]: Fair enough, thanks for the feedback!

File: source/grid/grid_generator_from_name.cc [hunk 9]
```diff
@@ -0,0 +1,331 @@
...
+    void
+    parse_and_create(void (*generator)(Triangulation<dim, spacedim> &,
+                                       Arguments...),
+                     const std::string &           arguments,
+                     Triangulation<dim, spacedim> &tria)
+    {
+      std::function<void(Arguments...)> wrapper =
+        [&tria, &generator](Arguments... args) { generator(tria, args...); };
+      auto bound_function = Utilities::mutable_bind(wrapper);
+      bound_function.parse_arguments(arguments);
```
[the remark this answers is not in the document]
Comment (luca-heltai) [reply source/grid/grid_generator_from_name.cc]: Thanks! :)

**Comment by luca-heltai** [comment]
ping @dealii/dealii

**Pull request merged**